// bitpay-sdk/src/Model/Bill/Bill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bill = void 0;
var Currency_1 = require("../../Currency");
var BitPayException_1 = require("../../Exceptions/BitPayException");
var Item_1 = require("./Item");
var Bill = /** @class */ (function () {
    /**
     * Constructor, create a minimal request Bill object.
     *
     * @param number   A string for tracking purposes.
     * @param currency The three digit currency type used to compute the bill's amount.
     * @param email    The email address of the receiver for this bill.
     * @param items    The list of itens to add to this bill.
     */
    function Bill(number, currency, email, items) {
        this.number = number;
        this.setCurrency(currency);
        this.email = email;
        this.items = items;
    }
    Bill.prototype.setCurrency = function (_currency) {
        if (!Currency_1.Currency.isValid(_currency))
            throw new BitPayException_1.default(null, "Error: currency code must be a type of Model.Currency", null);
        this.currency = _currency;
    };
    return Bill;
}());
exports.Bill = Bill;
exports.default = Item_1.Item;
