// bitpay-sdk/src/Exceptions/index.js
"use strict";
/*
 *     __    _ __
 *    / /_  (_) /_____  ____ ___  __
 *   / __ \/ / __/ __ \/ __ `/ / / /
 *  / /_/ / / /_/ /_/ / /_/ / /_/ /
 * /_.___/_/\__/ .___/\__,_/\__, /
 *            /_/          /____/
 *
 * BitPay NodeJS Client
 *
 * Copyright (c) 2020 BitPay inc.
 * This file is open source and available under the MIT license.
 * See the LICENSE file for more info.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionUpdate = exports.SubscriptionQuery = exports.SubscriptionCreation = exports.SubscriptionGeneric = exports.SettlementQuery = exports.SettlementGeneric = exports.PayoutDelete = exports.PayoutUpdate = exports.PayoutCancellation = exports.PayoutQuery = exports.PayoutCreation = exports.PayoutGeneric = exports.LedgerQuery = exports.LedgerGeneric = exports.BillUpdate = exports.BillDelivery = exports.BillQuery = exports.BillCreation = exports.BillGeneric = exports.RefundCancellation = exports.RefundQuery = exports.RefundCreation = exports.RefundGeneric = exports.RateQuery = exports.RateGeneric = exports.InvoiceQuery = exports.InvoiceCreation = exports.InvoiceGeneric = exports.Generic = void 0;
var BitPayException_1 = require("./BitPayException");
Object.defineProperty(exports, "Generic", { enumerable: true, get: function () { return BitPayException_1.BitPayException; } });
var InvoiceException_1 = require("./InvoiceException");
Object.defineProperty(exports, "InvoiceGeneric", { enumerable: true, get: function () { return InvoiceException_1.InvoiceException; } });
var InvoiceCreationException_1 = require("./InvoiceCreationException");
Object.defineProperty(exports, "InvoiceCreation", { enumerable: true, get: function () { return InvoiceCreationException_1.InvoiceCreationException; } });
var InvoiceQueryException_1 = require("./InvoiceQueryException");
Object.defineProperty(exports, "InvoiceQuery", { enumerable: true, get: function () { return InvoiceQueryException_1.InvoiceQueryException; } });
var RateException_1 = require("./RateException");
Object.defineProperty(exports, "RateGeneric", { enumerable: true, get: function () { return RateException_1.RateException; } });
var RateQueryException_1 = require("./RateQueryException");
Object.defineProperty(exports, "RateQuery", { enumerable: true, get: function () { return RateQueryException_1.RateQueryException; } });
var RefundException_1 = require("./RefundException");
Object.defineProperty(exports, "RefundGeneric", { enumerable: true, get: function () { return RefundException_1.RefundException; } });
var RefundCreationException_1 = require("./RefundCreationException");
Object.defineProperty(exports, "RefundCreation", { enumerable: true, get: function () { return RefundCreationException_1.RefundCreationException; } });
var RefundQueryException_1 = require("./RefundQueryException");
Object.defineProperty(exports, "RefundQuery", { enumerable: true, get: function () { return RefundQueryException_1.RefundQueryException; } });
var RefundCancellationException_1 = require("./RefundCancellationException");
Object.defineProperty(exports, "RefundCancellation", { enumerable: true, get: function () { return RefundCancellationException_1.RefundCancellationException; } });
var BillException_1 = require("./BillException");
Object.defineProperty(exports, "BillGeneric", { enumerable: true, get: function () { return BillException_1.BillException; } });
var BillCreationException_1 = require("./BillCreationException");
Object.defineProperty(exports, "BillCreation", { enumerable: true, get: function () { return BillCreationException_1.BillCreationException; } });
var BillQueryException_1 = require("./BillQueryException");
Object.defineProperty(exports, "BillQuery", { enumerable: true, get: function () { return BillQueryException_1.BillQueryException; } });
var BillDeliveryException_1 = require("./BillDeliveryException");
Object.defineProperty(exports, "BillDelivery", { enumerable: true, get: function () { return BillDeliveryException_1.BillDeliveryException; } });
var BillUpdateException_1 = require("./BillUpdateException");
Object.defineProperty(exports, "BillUpdate", { enumerable: true, get: function () { return BillUpdateException_1.BillUpdateException; } });
var LedgerException_1 = require("./LedgerException");
Object.defineProperty(exports, "LedgerGeneric", { enumerable: true, get: function () { return LedgerException_1.LedgerException; } });
var LedgerQueryException_1 = require("./LedgerQueryException");
Object.defineProperty(exports, "LedgerQuery", { enumerable: true, get: function () { return LedgerQueryException_1.LedgerQueryException; } });
var PayoutException_1 = require("./PayoutException");
Object.defineProperty(exports, "PayoutGeneric", { enumerable: true, get: function () { return PayoutException_1.PayoutException; } });
var PayoutCreationException_1 = require("./PayoutCreationException");
Object.defineProperty(exports, "PayoutCreation", { enumerable: true, get: function () { return PayoutCreationException_1.PayoutCreationException; } });
var PayoutQueryException_1 = require("./PayoutQueryException");
Object.defineProperty(exports, "PayoutQuery", { enumerable: true, get: function () { return PayoutQueryException_1.PayoutQueryException; } });
var PayoutCancellationException_1 = require("./PayoutCancellationException");
Object.defineProperty(exports, "PayoutCancellation", { enumerable: true, get: function () { return PayoutCancellationException_1.PayoutCancellationException; } });
var PayoutUpdateException_1 = require("./PayoutUpdateException");
Object.defineProperty(exports, "PayoutUpdate", { enumerable: true, get: function () { return PayoutUpdateException_1.PayoutUpdateException; } });
var PayoutDeleteException_1 = require("./PayoutDeleteException");
Object.defineProperty(exports, "PayoutDelete", { enumerable: true, get: function () { return PayoutDeleteException_1.PayoutDeleteException; } });
var SettlementException_1 = require("./SettlementException");
Object.defineProperty(exports, "SettlementGeneric", { enumerable: true, get: function () { return SettlementException_1.SettlementException; } });
var SettlementQueryException_1 = require("./SettlementQueryException");
Object.defineProperty(exports, "SettlementQuery", { enumerable: true, get: function () { return SettlementQueryException_1.SettlementQueryException; } });
var SubscriptionException_1 = require("./SubscriptionException");
Object.defineProperty(exports, "SubscriptionGeneric", { enumerable: true, get: function () { return SubscriptionException_1.SubscriptionException; } });
var SubscriptionCreationException_1 = require("./SubscriptionCreationException");
Object.defineProperty(exports, "SubscriptionCreation", { enumerable: true, get: function () { return SubscriptionCreationException_1.SubscriptionCreationException; } });
var SubscriptionQueryException_1 = require("./SubscriptionQueryException");
Object.defineProperty(exports, "SubscriptionQuery", { enumerable: true, get: function () { return SubscriptionQueryException_1.SubscriptionQueryException; } });
var SubscriptionUpdateException_1 = require("./SubscriptionUpdateException");
Object.defineProperty(exports, "SubscriptionUpdate", { enumerable: true, get: function () { return SubscriptionUpdateException_1.SubscriptionUpdateException; } });
